### PR TITLE
Add native iOS screenshots to integration_test

### PR DESCRIPTION
**Pull request by jmagman**
Implement the native iOS screenshot side of https://github.com/flutter/flutter/pull/84472.  Added an optional "name" argument for the "captureScreenshot" method.

When run as a native Xcode test, the screenshots are saved to the Xcode test results bundle in addition to the bytes being passed back over the channel.
Screen Shot 2021-07-22 at 4 32 46 PM

Screenshot is the app window, not including the status bar, etc.
public png_1_7EACEC54-4CF2-42E1-A6B3-D203D265DD1E

iOS screenshot prototype of https://github.com/flutter/flutter/issues/83856

### CODE REVIEW

File: packages/integration_test/lib/integration_test.dart [hunk 1]
```diff
@@ -171,7 +173,15 @@ https://flutter.dev/docs/testing/integration-tests#testing-on-firebase-test-lab
   ///
   /// Called by test methods. Implementation differs for each platform.
   Future<void> takeScreenshot(String screenshotName) async {
-    await callbackManager.takeScreenshot(screenshotName);
+    try {
```
Comment (jmagman) [inline packages/integration_test/lib/integration_test.dart:176]: This should not be checked in.

File: dev/integration_tests/ui/ios/Runner.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jmagman) [inline dev/integration_tests/ui/ios/Runner.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:1]: This should have been checked in before.

File: packages/integration_test/example/ios/Runner.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -475,21 +475,11 @@
 			baseConfigurationReference = 09505407E99803EF7AA92DE7 /* Pods-RunnerTests.debug.xcconfig */;
 			buildSettings = {
 				BUNDLE_LOADER = "$(TEST_HOST)";
-				CLANG_ANALYZER_NUMBER_OBJECT_CONVERSION = YES_AGGRESSIVE;
```
Comment (jmagman) [inline packages/integration_test/example/ios/Runner.xcodeproj/project.pbxproj:478]: Remove unnecessary build setting overrides.  I can do this in a separate PR, if you prefer.

File: packages/integration_test/ios/Classes/IntegrationTestIosTest.h [hunk 4]
```diff
@@ -4,23 +4,40 @@
...
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol FLTIntegrationTestScreenshotDelegate;
+
 @interface IntegrationTestIosTest : NSObject
 
-- (BOOL)testIntegrationTest:(NSString **)testResult;
+- (instancetype)initWithScreenshotDelegate:(nullable id<FLTIntegrationTestScreenshotDelegate>)delegate NS_DESIGNATED_INITIALIZER;
+
+- (BOOL)testIntegrationTest:(NSString *_Nullable *_Nullable)testResult;
```
Comment (stuartmorgan) [inline packages/integration_test/ios/Classes/IntegrationTestIosTest.h:15]: Could we get a comment on this weird API describing its behavior while you are touching it? (Or even better, could we make this an idiomatic ObjC method that returns either an NSString or nil?)

File: packages/integration_test/ios/Classes/IntegrationTestPlugin.m [hunk 5]
```diff
@@ -39,20 +44,53 @@ + (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
...
+  } else if ([call.method isEqualToString:kMethodConvertSurfaceToImage] || [call.method isEqualToString:kMethodRevertImage]) {
+    // Android only, no-op on iOS.
+    result(nil);
   } else {
     result(FlutterMethodNotImplemented);
   }
 }
 
+- (UIImage *)capturePngScreenshot {
+  UIWindow *window = [UIApplication.sharedApplication.windows filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"keyWindow = YES"]].firstObject;
```
Comment (stuartmorgan) [inline packages/integration_test/ios/Classes/IntegrationTestPlugin.m:76]: Looks like we're not running clang-format in this directory on CI? This is way over the 100 char style guide length limit.

File: packages/integration_test/lib/_callback_io.dart [hunk 6]
```diff
@@ -60,8 +61,8 @@ class IOCallbackManager implements CallbackManager {
     // comes up in the future. For example: `WebCallbackManager.cleanup`.
   }
 
-  // Whether the Flutter surface uses an Image.
-  bool _usesFlutterImage = false;
+  // Whether the Flutter surface uses an Image (Android only).
```
Comment (stuartmorgan) [inline packages/integration_test/lib/_callback_io.dart:64]: This sounds to me like using an image is Android-only, but the logic below is the reverse.

**Review by stuartmorgan-g — APPROVED** [review]
LGTM with nits. Shouldn't we have a test of this functionality that screenshots a dummy plugin and compares it to a golden image though?

File: packages/integration_test/lib/_callback_io.dart (hunk 6 above)
Comment (jmagman) [reply packages/integration_test/lib/_callback_io.dart:64]: @blasten What does `Whether the Flutter surface uses an Image` mean in this context?  I think that the surface is frozen, the screenshot taken, and then it's unfrozen on Android?

Comment (blasten) [reply packages/integration_test/lib/_callback_io.dart:64]: oh - yes. Image really means that Flutter is using this class https://github.com/flutter/engine/blob/master/shell/platform/android/io/flutter/embedding/android/FlutterImageView.java. 👻

File: packages/integration_test/lib/_callback_io.dart [hunk 7]
```diff
@@ -90,7 +91,7 @@ class IOCallbackManager implements CallbackManager {
     integrationTestChannel.setMethodCallHandler(_onMethodChannelCall);
     final List<int>? rawBytes = await integrationTestChannel.invokeMethod<List<int>>(
       'captureScreenshot',
-      null,
+      <String, dynamic>{'name': screenshot},
```
Comment (blasten) [inline packages/integration_test/lib/_callback_io.dart:94]: the above `!_usesFlutterImage` if statement could also check if `Platform.isAndroid`

File: packages/integration_test/lib/_callback_io.dart [hunk 8]
```diff
@@ -60,8 +61,8 @@ class IOCallbackManager implements CallbackManager {
     // comes up in the future. For example: `WebCallbackManager.cleanup`.
   }
 
-  // Whether the Flutter surface uses an Image.
-  bool _usesFlutterImage = false;
+  // Whether the Flutter surface uses an Image (Android only).
+  bool _usesFlutterImage = false || !Platform.isAndroid;
 
   @override
   Future<void> convertFlutterSurfaceToImage() async {
```
Comment (blasten) [inline packages/integration_test/lib/_callback_io.dart:68]: it would need to noop if `!Platform.isAndroid`

**Review by blasten — APPROVED** [review]
LGTM % nits

File: packages/integration_test/ios/Classes/IntegrationTestIosTest.h (hunk 4 above)
Comment (jmagman) [reply packages/integration_test/ios/Classes/IntegrationTestIosTest.h:15]: This API is very weird (I just added the nullability), and the `INTEGRATION_TEST_IOS_RUNNER` pragma is downright Swift-hostile...

File: packages/integration_test/ios/Classes/IntegrationTestPlugin.m (hunk 5 above)
Comment (jmagman) [reply packages/integration_test/ios/Classes/IntegrationTestPlugin.m:76]: I don't think there are any formatters (even dartfmt) running on flutter/flutter (plus I hate how clang formatter mangles Obj-C blocks).  Let me shorten these up.

File: packages/integration_test/lib/_callback_io.dart (hunk 8 above)
Comment (jmagman) [reply packages/integration_test/lib/_callback_io.dart:68]: I meant to make it so it was never set to false on `!Platform.isAndroid`, let me clean this up.

**Review by jmagman — COMMENTED** [review]
> LGTM with nits. Shouldn't we have a test of this functionality that screenshots a dummy plugin and compares it to a golden image though?

I think the next step for @blasten after I commit this was to set up the screenshot integration tests.  But I should also write some iOS native tests.

**Comment by jmagman** [comment]
> LGTM with nits. Shouldn't we have a test of this functionality that screenshots a dummy plugin and compares it to a golden image though?

There will be some test harness work needed for the native side.  Filed https://github.com/flutter/flutter/issues/87770.

We'll get some unit tests for this going soon.

**Comment by blasten** [comment]
🎉

**Pull request merged**